visualizer/semanticsHelper.js

```diff
@@ -259,12 +289,82 @@
     return pexpr;
   }
 
+  // Build the action that called from the wrapper, which includes renaming the
+  // operation arguments, and actual action body that user typed.
+  function buildAction(opArgs, enclosedActionArgStr, actionBody) {
+    // Constructs the code that renames the operation arguments, so we could refer
+    // them by name directly. i.e. `var <argument name> = this.args.<argument name>`
+    var renameOpArgStr = opArgs &&  Object.keys(opArgs).map(function(argName) {
+      return 'var ' + argName + ' = this.args.' + argName + ';';
+    }).join('\n');
+
+    var actionStr = 'function' + enclosedActionArgStr + '{\n' +
+        renameOpArgStr + '\n' + actionBody +
+      '\n}';
+    return eval('(' + actionStr + ')');    // eslint-disable-line no-eval
+  }
+
+  function wrapAction(currentOpName, enclosedActionArgStr, actionBody) {
+    if (!actionBody.trim()) {
+      return undefined;
+    }
+
+    var realAction = 'function' + enclosedActionArgStr + '{\n' + actionBody + '\n}';
+    var isExpression = es6.match(actionBody, 'AssignmentExpression<withIn>').succeeded();
+    if (isExpression) {
+      actionBody = 'return ' + actionBody + ';';
+    }
+    var wrapper = function() {
+      var key = nodeKey(this);
+      var result;
+      var origTodoList = todoList;
+      todoList = undefined;
+      try {
+        var action = buildAction(this.args, enclosedActionArgStr, actionBody);
+        result = action.apply(this, arguments);
```

Comment:
...and then I would put this line in a separate try/catch as well, because there is a completely different type of error that might get thrown here. Then I think you might be able to get rid of the `isErrorWrapper` check below.